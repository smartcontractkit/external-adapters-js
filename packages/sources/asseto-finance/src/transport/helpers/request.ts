export interface ApiRequestConfig {
  baseURL: string
  url?: string
  headers?: Record<string, string>
  method?: string
  data?: any
}

export class RequestHelper {
  static createAuthenticatedGetRequest(
    baseURL: string,
    bearerToken: string,
    path?: string,
  ): ApiRequestConfig {
    return {
      baseURL,
      url: path,
      headers: {
        Authorization: `Bearer ${bearerToken}`,
      },
    }
  }

  static createNavRequest(baseURL: string, bearerToken: string, fundId: number): ApiRequestConfig {
    return this.createAuthenticatedGetRequest(baseURL, bearerToken, `/${fundId}/nav-daily`)
  }

  static createReserveRequest(
    baseURL: string,
    bearerToken: string,
    fundId: number,
  ): ApiRequestConfig {
    return this.createAuthenticatedGetRequest(baseURL, bearerToken, `/${fundId}/reserves`)
  }
}
