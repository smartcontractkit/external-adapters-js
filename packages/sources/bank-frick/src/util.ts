import { makeLogger } from '@chainlink/external-adapter-framework/util'
import axios, { AxiosRequestConfig } from 'axios'
import crypto from 'crypto'
import { config } from './config'
import { SigningAlgorithm } from './types'

interface TokenResponseBody {
  errors?: string[]
  token: string
}
interface TokenRequestBody {
  key: string
}

const logger = makeLogger('BankFrickUtil')

// Used by all endpoints requiring authentication
export const generateJWT = async (
  settings: typeof config.settings,
  signingAlgorithm: SigningAlgorithm = 'rsa-sha512',
): Promise<string> => {
  logger.info("Generating a new JWT because we don't have one in config.token")
  const { API_KEY, PRIVATE_KEY, API_ENDPOINT } = settings

  // All of these are required, so validation should have failed prior to this line
  if (!API_KEY || !PRIVATE_KEY) {
    throw new Error(
      'API_KEY, PRIVATE_KEY, and PASSWORD all must be defined to get a new token\n' +
        'Received:  \n' +
        `API_KEY: ${API_KEY}\n` +
        `PRIVATE_KEY: ${PRIVATE_KEY}`,
    )
  }
  let privateKey = PRIVATE_KEY

  if (!privateKey.match(/-----?BEGIN ([A-Z ])*PRIVATE KEY-----?/)) {
    privateKey = Buffer.from(privateKey, 'base64').toString('utf8')
  }

  const data: TokenRequestBody = {
    key: API_KEY,
  }

  const signature = crypto.sign(signingAlgorithm, Buffer.from(JSON.stringify(data)), privateKey)
  const options: AxiosRequestConfig<TokenRequestBody> = {
    method: 'POST',
    baseURL: API_ENDPOINT,
    url: `authorize`,
    headers: {
      Signature: signature.toString('base64'),
      algorithm: signingAlgorithm,
    },
    data,
  }

  const response = await axios.request<TokenResponseBody>(options)

  return response.data.token
}
