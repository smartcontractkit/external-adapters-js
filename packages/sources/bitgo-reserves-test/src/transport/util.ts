import * as crypto from 'crypto'

export interface DataSchema {
  totalReserve: string
  cashReserve: string
  investedReserve: string
  lastUpdated: string
}

export interface ResponseSchema {
  data: string // formatted & escaped DataSchema
  dataSignature: string
  ripcord: boolean | string
}

export function createRequest(params: { [key: string]: string }[], endpoint: string) {
  return params.map((param) => {
    return {
      params: [param],
      request: {
        baseURL: endpoint,
      },
    }
  })
}

export function parseResponse(
  params: { [key: string]: string }[],
  payload: ResponseSchema,
  pubkey: string,
) {
  if (!payload || !payload.data) {
    return params.map((param) => {
      return {
        params: param,
        response: {
          errorMessage: `The data provider didn't return any value`,
          statusCode: 502,
        },
      }
    })
  }

  if (
    (typeof payload.ripcord === 'boolean' && payload.ripcord) ||
    (typeof payload.ripcord === 'string' && payload.ripcord.toUpperCase() === 'TRUE')
  ) {
    return [
      {
        params: params[0],
        response: {
          errorMessage: 'Ripcord indicator true',
          ripcord: payload.ripcord,
          statusCode: 502,
        },
      },
    ]
  }

  const verifier = crypto.createVerify('sha256')
  verifier.update(payload.data)
  if (!verifier.verify(pubkey, payload.dataSignature, 'base64')) {
    return params.map((param) => {
      return {
        params: param,
        response: {
          errorMessage: `Data verification failed`,
          statusCode: 502,
        },
      }
    })
  }

  const data = JSON.parse(payload.data) as DataSchema
  const timestamps = {
    providerIndicatedTimeUnixMs: new Date(data.lastUpdated).getTime(),
  }
  const result = Number(data.totalReserve)
  if (result === undefined || isNaN(result)) {
    return params.map((param) => {
      return {
        params: param,
        response: {
          errorMessage: `Missing totalReserve`,
          statusCode: 502,
        },
      }
    })
  }

  return params.map((param) => {
    return {
      params: param,
      response: {
        result,
        data: {
          result,
        },
        timestamps,
      },
    }
  })
}

export function normalizePubkey(envPubkey: string): string {
  return envPubkey.replace(/\\n/g, '\n')
}
