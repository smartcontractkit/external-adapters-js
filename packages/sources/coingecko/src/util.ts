import { makeExecute } from './adapter'
import { executeSync } from '@chainlink/ea-bootstrap'
import { CoinsResponse } from './endpoint/coins'

export function getCoinIds(id: string): Promise<CoinsResponse[]> {
  const execute = makeExecute()
  const executeWithMiddleware = executeSync(execute)
  const options = {
    data: {
      endpoint: 'coins',
      maxAge: 60 * 60 * 1000, // 1 hour
    },
    method: 'post',
    id,
  }
  return new Promise((resolve, reject) => {
    executeWithMiddleware(options, (_, result) => {
      if (result.error) {
        reject(result)
      } else {
        resolve(result.data)
      }
    })
  })
}

export const getSymbolsToIds = (
  symbols: string[],
  coinList: CoinsResponse[],
): Record<string, string> => {
  const idToSymbol: Record<string, string> = {}
  symbols.forEach((symbol) => {
    const byId = coinList.find((d) => d.id.toLowerCase() === symbol.toLowerCase())
    if (byId) idToSymbol[byId.id] = byId.symbol
    const coin = coinList.find((d) => d.symbol.toLowerCase() === symbol.toLowerCase())
    if (coin && coin.id) {
      idToSymbol[coin.id] = symbol
    }
  })
  return idToSymbol
}
