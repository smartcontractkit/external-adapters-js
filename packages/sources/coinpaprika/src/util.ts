import { makeExecute } from './adapter'
import { makeMiddleware, withMiddleware } from '@chainlink/ea-bootstrap'
import { CoinsResponse } from './endpoint/coins'
import { AdapterContext } from '@chainlink/types'

export function getCoinIds(context: AdapterContext, id: string): Promise<CoinsResponse[]> {
  const execute = makeExecute()
  const options = {
    data: {
      endpoint: 'coins',
      maxAge: 60 * 60 * 1000, // 1 hour
    },
    method: 'post',
    id,
  }
  return new Promise((resolve, reject) => {
    const middleware = makeMiddleware(execute)
    withMiddleware(execute, context, middleware)
      .then((executeWithMiddleware) => {
        executeWithMiddleware(options, context).then((value) => resolve(value.data))
      })
      .catch((error) => reject(error))
  })
}

export const getSymbolToId = (symbol: string, coinList: CoinsResponse[]): string => {
  const isId = coinList.find(({ id }) => id.toLowerCase() === symbol.toLowerCase())
  if (isId && isId.id) return isId.id.toLowerCase()

  const coin = coinList.find(
    ({ symbol: coinSymbol, rank }) =>
      coinSymbol.toLowerCase() === symbol.toLowerCase() && rank !== 0,
  )
  if (coin && coin.id) return coin.id.toLowerCase()

  throw new Error('Coin id not found')
}
