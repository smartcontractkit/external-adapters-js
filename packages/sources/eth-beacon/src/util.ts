import { endpointSelector, makeExecute } from './adapter'
import {
  AdapterData,
  AdapterResponseData,
  makeMiddleware,
  withMiddleware,
} from '@chainlink/ea-bootstrap'
import { AdapterContext, Config } from '@chainlink/ea-bootstrap'
import * as endpoints from './endpoint'

export function requestSelf(
  context: AdapterContext,
  id: string,
  input: endpoints.TInputParameters & { endpoint: string },
): Promise<AdapterResponseData<AdapterData>> {
  const execute = makeExecute()
  const options = {
    data: input,
    method: 'post',
    id,
  }
  return new Promise((resolve, reject) => {
    const middleware = makeMiddleware<Config, endpoints.TInputParameters>(
      execute,
      undefined,
      endpointSelector,
    )
    withMiddleware<endpoints.TInputParameters>(execute, context, middleware)
      .then((executeWithMiddleware) => {
        executeWithMiddleware(options, context)
          .then((value) => resolve(value.data))
          .catch(reject)
      })
      .catch((error) => reject(error))
  })
}
