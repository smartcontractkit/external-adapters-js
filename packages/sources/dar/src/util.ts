import { makeLogger } from '@chainlink/external-adapter-framework/util'
import {
  AdapterConnectionError,
  AdapterDataProviderError,
  AdapterError,
} from '@chainlink/external-adapter-framework/validation/error'
import axios from 'axios'
import { config } from './config'
import { AuthResponse } from './types'

const logger = makeLogger('DarUtil')

export const getAuthToken = async (settings: typeof config.settings): Promise<string> => {
  const requestedTs = Date.now()
  try {
    const buf = Buffer.from(`${settings.WS_API_USERNAME}:${settings.WS_API_KEY}`)
    const auth = buf.toString('base64')
    const jwtRes = await axios.request<AuthResponse>({
      url: `${settings.API_ENDPOINT}/token-auth`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        Authorization: `Basic ${auth}`,
      },
    })

    return jwtRes.data.access_token
  } catch (e: unknown) {
    const err = e as any
    const message = `Login failed ${err.message ? `with message '${err.message}'` : ''}`
    const error = { ...err, message }

    logger.debug(message)

    throw error.response
      ? new AdapterDataProviderError(error, {
          providerDataRequestedUnixMs: requestedTs,
          providerIndicatedTimeUnixMs: undefined,
          providerDataReceivedUnixMs: Date.now(),
        })
      : error.request
      ? new AdapterConnectionError(error, {
          providerDataRequestedUnixMs: requestedTs,
          providerIndicatedTimeUnixMs: undefined,
          providerDataReceivedUnixMs: Date.now(),
        })
      : new AdapterError(error)
  }
}
