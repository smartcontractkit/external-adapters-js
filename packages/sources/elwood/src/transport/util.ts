import axios from 'axios'
import { makeLogger } from '@chainlink/external-adapter-framework/util'

export type SubscribeRequest = {
  action: 'subscribe' | 'unsubscribe'
  stream: 'index'
  symbol: string
  index_freq: number
}

export type PriceResponse = {
  type: 'Index'
  data: {
    price: string
    bid: string
    ask: string
    symbol: string
    timestamp: string
  }
  sequence: number
}

export type HeartbeatResponse = {
  type: 'heartbeat'
  data: string
  sequence: number
}

export type ErrorResponse = {
  type: unknown
  error: Record<string, unknown>
}

export type ResponseMessage = PriceResponse | HeartbeatResponse

export type LoggerType = ReturnType<typeof makeLogger>

export interface ValidatedWsMessage {
  base: string
  quote: string
  result: number
  bid: number
  ask: number
  timestamp: number
}

export const buildUrl = (endpoint: string, apiKey: string) => `${endpoint}?apiKey=${apiKey}`

export const validateWsMessage = (
  logger: LoggerType,
  message: ResponseMessage,
): ValidatedWsMessage | null => {
  if (message.type !== 'Index') {
    return null
  }

  if (!message.data) {
    logger.warn(`Got no data in WS message of type Index`)
    return null
  }

  if (typeof message.data?.symbol !== 'string') {
    logger.warn(`Got non string symbol "${message.data?.symbol}" in WS message of type Index`)
    return null
  }

  const [base, quote] = message.data.symbol.split('-')
  if (!base || !quote) {
    logger.warn(`Got invalid symbol "${message.data?.symbol}" in WS message of type Index`)
    return null
  }

  const result = Number(message.data.price)
  if (result < 0) {
    logger.warn(`Got invalid price "${message.data.price}" in WS message of type Index`)
    return null
  }

  return {
    base,
    quote,
    result,
    bid: Number(message.data.bid),
    ask: Number(message.data.ask),
    timestamp: new Date(message.data.timestamp).getTime(),
  }
}

export const buildWsMessage = (
  action: 'subscribe' | 'unsubscribe',
  params: { base: string; quote: string },
): SubscribeRequest => ({
  action,
  stream: 'index',
  symbol: `${params.base}-${params.quote}`,
  index_freq: 1_000, // Milliseconds
})

export const sendMessage = async (endpoint: string, apiKey: string, message: SubscribeRequest) =>
  axios.request({
    url: buildUrl(endpoint, apiKey),
    method: 'post',
    data: message,
  })

type Subscription = {
  index_freq: number
  stream: string
  symbol: string
}

type SubscriptionsSuccessResponse = {
  data: {
    items: (Subscription | string)[]
  }
}

type SubscriptionsResponse = SubscriptionsSuccessResponse | ErrorResponse

export const getSubscriptionKey = (subscription: Subscription): string =>
  `${subscription.stream}:${subscription.symbol}:${subscription.index_freq}`

export const getSubscriptions = async (
  endpoint: string,
  apiKey: string,
  logger: LoggerType,
): Promise<Set<string>> => {
  let data: SubscriptionsResponse
  try {
    const response = await axios.get<SubscriptionsResponse>(
      buildUrl(`${endpoint}/subscriptions`, apiKey),
    )
    data = response.data
  } catch (e) {
    logger.warn(`Failed to get current subscriptions, assuming no subscriptions: ${e}`)
    return new Set()
  }

  if ('error' in data) {
    logger.warn(
      `Subscriptions response has error, assuming no subscriptions: ${JSON.stringify(data.error)}`,
    )
    return new Set()
  }

  const subscriptions: string[] = []
  data.data.items.forEach((item) => {
    if (typeof item === 'object') {
      subscriptions.push(getSubscriptionKey(item))
    }
  })
  logger.info(`Currently subscribed to: ${subscriptions}`)

  return new Set(subscriptions)
}
