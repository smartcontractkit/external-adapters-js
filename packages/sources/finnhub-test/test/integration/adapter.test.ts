import * as process from 'process'
import { AddressInfo } from 'net'
import request, { SuperTest, Test } from 'supertest'
import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { AdapterRequestBody, sleep } from '@chainlink/external-adapter-framework/util'
import { mockResponseSuccess } from './fixtures'
import { setEnvVariables, createAdapter } from './setup'

describe('rest', () => {
  jest.setTimeout(10000)
  let spy: jest.SpyInstance
  beforeAll(async () => {
    const mockDate = new Date('2022-01-01T11:11:11.111Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())
  })

  afterAll((done) => {
    spy.mockRestore()
    done()
  })

  let fastify: ServerInstance | undefined
  let req: SuperTest<Test>

  const data: AdapterRequestBody = {
    data: {
      base: 'EUR',
    },
  }

  let oldEnv: NodeJS.ProcessEnv
  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env['CACHE_MAX_AGE'] = '5000'
    process.env['CACHE_POLLING_MAX_RETRIES'] = '0'
    process.env['METRICS_ENABLED'] = 'false'
    process.env['API_KEY'] = 'fake-api-key'
    fastify = await expose(createAdapter())
    req = request(`http://localhost:${(fastify?.server.address() as AddressInfo).port}`)
    mockResponseSuccess()
    // Send initial request to start background execute
    await req.post('/').send(data)
    await sleep(5000)
  })

  afterAll((done) => {
    setEnvVariables(oldEnv)
    fastify?.close(done())
  })

  describe('quote endpoint', () => {
    it('should return success', async () => {
      const makeRequest = () =>
        req
          .post('/')
          .send(data)
          .set('Accept', '*/*')
          .set('Content-Type', 'application/json')
          .expect('Content-Type', /json/)

      const response = await makeRequest()
      expect(response.body).toMatchSnapshot()
    }, 30000)
  })
})
