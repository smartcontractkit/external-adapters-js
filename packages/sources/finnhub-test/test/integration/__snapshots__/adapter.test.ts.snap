// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest quote endpoint should return success 1`] = `
{
  "data": {
    "result": 1.15894,
  },
  "result": 1.15894,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1641035471111,
    "providerDataRequestedUnixMs": 1641035471111,
  },
}
`;
